go/store/prolly/sort/external.go

```diff
@@ -0,0 +1,501 @@
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"sync"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/dolthub/dolt/go/store/util/tempfiles"
+	"github.com/dolthub/dolt/go/store/val"
+)
+
+// tupleSorter inputs a series of unsorted tuples and outputs a sorted list
+// of tuples. Batches of tuples sorted in memory are written to disk, and
+// then k-way merge sorted to produce a final sorted list. The |fileMax|
+// parameter limits the number of files spilled to disk at any given time.
+// The maximum memory used will be |fileMax| * |batchSize|.
+type tupleSorter struct {
+	keyCmp    func(val.Tuple, val.Tuple) bool
+	files     []keyIterable
+	inProg    *keyMem
+	fileMax   int
+	batchSize int
+	tmpProv   tempfiles.TempFileProvider
+}
+
+func NewTupleSorter(batchSize, fileMax int, keyCmp func(val.Tuple, val.Tuple) bool, tmpProv tempfiles.TempFileProvider) *tupleSorter {
+	if fileMax%2 == 1 {
+		// round down to even
+		// fileMax/2 will be compact parallelism
+		fileMax -= 1
+	}
+	ret := &tupleSorter{
+		fileMax:   fileMax,
+		batchSize: batchSize,
+		keyCmp:    keyCmp,
+		tmpProv:   tmpProv,
+	}
+	ret.inProg = newKeyMem(ret.newFile(), batchSize)
+	return ret
+}
+
+func (a *tupleSorter) Flush(ctx context.Context) (iter keyIterable, err error) {
+	defer func() {
+		err = recoverErr()
+	}()
+
+	// don't flush in-progress, just sort in memory
+	a.inProg.sort(a.keyCmp)
+
+	if len(a.files) == 0 {
+		// don't go to disk if we didn't reach a mem flush
+		return a.inProg, nil
+	}
+
+	var iterables []keyIterable
+	iterables = append(iterables, a.inProg)
+	for _, file := range a.files {
+		iterables = append(iterables, file)
+	}
+
+	allKeys := newKeyFile(a.newFile(), a.inProg.byteLim)
+	m := newFileMerger(ctx, a.keyCmp, allKeys, iterables...)
+	m.run(ctx)
+	return allKeys, nil
+}
+
+func (a *tupleSorter) Insert(ctx context.Context, k val.Tuple) (err error) {
+	defer func() {
+		err = recoverErr()
+	}()
+
+	if !a.inProg.insert(k) {
+		a.flushMem(ctx)
+		a.inProg.insert(k)
+	}
+	return
+}
+
+func (a *tupleSorter) flushMem(ctx context.Context) {
+	// flush and replace |inProg|
+	if a.inProg.Len() > 0 {
+		newFile := a.inProg.flush(a.keyCmp)
+		a.inProg = newKeyMem(a.newFile(), a.batchSize)
+		a.files = append(a.files, newFile)
+	}
+	if len(a.files) >= a.fileMax {
+		// merge sort files
+		a.compact(ctx)
+	}
+}
+
+func (a *tupleSorter) newFile() *os.File {
+	f, err := a.tmpProv.NewFile("", "key_file_")
+	if err != nil {
+		newError(err)
+	}
+	return f
+}
+
+// compact halves the number of files, doubling their size
+func (a *tupleSorter) compact(ctx context.Context) {
+	var newFiles []keyIterable
+	wg := sync.WaitGroup{}
+	eg, ctx := errgroup.WithContext(ctx)
+	work := make(chan *keyFile)
+	eg.Go(func() error {
+		for {
+			select {
+			case newF, ok := <-work:
+				if !ok {
+					return nil
+				}
+				newFiles = append(newFiles, newF)
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}
+	})
+	for i := 0; i < len(a.files); i += 2 {
```

Comment:
If I understand the whole logic here, the asymptotics on this strategy as implemented here seem off to me, at least compared to optimality. AFAIU:

1) we're scanning the entire data set on every compact, and
2) every compact is coming at fixed size intervals, for example, every time we have 64 new files of 32MBs, we scan every existing file...

Don't we get N^2 asymptotics?

If you take a look at the file sizes of the files in each slot, you'll also see we're just building one big file at index 0 and everything else stays quite a bit smaller than it...so it's also probably not great for parallelism, because we're going to block on the linear scan of that file each compact... 

I think an approach where we only merge the n/2 smallest files each `compact` (we only get by `n/4` slots in that case) gets us back to better asymptotics, but there are probably much better approaches which take advantage of the k-way merge (instead of 2-way...)